OCPBUGS-32321: Add username/uid label and annotation for the user setting ConfigMap, Role and RoleBinding

jerolimov opened the pull request:
This change adds 3 new labels and 2 annotations for newly created user settings resources.

* Labels
  1. A new label that identifies this resource is a user-settings resource.
  2. A new label with the UID, if the user-settings refers a local OpenShift user, `kubeadmin`, or empty for OIDC users.
* Annotations:
  2. The origin username, as annotations also allow email addresses when an OIDC user is used.

@stlaz this should work for the different auth variants we support, or?

For example for a **kube:admin** ConfigMap:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-settings-kubeadmin
  labels:
    console.openshift.io/user-settings: "true"
    console.openshift.io/user-settings-uid: "kubeadmin"
  annotations:
    console.openshift.io/user-settings-username: "kube:admin"
```

For a **local User with a UID**:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-settings-1234-1234...
  labels:
    console.openshift.io/user-settings: "true"
    console.openshift.io/user-settings-uid: "1234-1234..."
  annotations:
    console.openshift.io/user-settings-username: "consoledeveloper"
```

For an **OIDC user without an UID but with an email as username**:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-settings-e3b0c442...
  labels:
    console.openshift.io/user-settings: "true"
    console.openshift.io/user-settings-uid: ""
  annotations:
    console.openshift.io/user-settings-username: "openshift@redhat.com"
```

/cc @stlaz @jhadvig @TheRealJon

jerolimov commented:
/jira refresh

jerolimov commented:
/retest

ranakan19 commented:
Hello 
Thank you for this PR, have a few questions:
1. When looking for resources created by the console, is one label enough to identify? I see that the label `console.openshift.io/user-settings: "true"` is present in all the examples you've mentioned and seems like a good option to use that.
2. Out of the three examples mentioned, it looks like resources created for users in developer sandbox would be of the second example - "developer with UID". If yes, this [PR](https://github.com/codeready-toolchain/member-operator/pull/568) uses the two labels `console.openshift.io/user-settings: "true"
 , console.openshift.io/user-settings-uid:` to identify the resources created by console for user, to clean it up, do you think these would be enough to identify them?
 3. Would it be possible to extend these labels to existing resources as well, maybe like a migration code? It would be easier to identify these resources instead of searching by name. Currently I've divided the code to look for resources by name first and then if not found look by labels - which isn't pretty.

jerolimov commented:
/retest

jerolimov commented:
/jira refresh

jerolimov commented:
Hi, @ranakan19. Thanks for your patience and first review. I've updated the PR. Can you take a fresh look?

/cc @stlaz as auth expert
/cc @jhadvig @TheRealJon as console backend experts
/cc @vikram-raj @lokanandaprabhu @Lucifergene as ODC reviewers
/cc @sanketpathak as ODC QE

Can some more people please take a look into that PR? Thanks :smile: 

I changed the PR so that it now creates only 2 labels and 1 annotation. See my updated initial comment.


I also created a small Node.js based script to apply the missing labels to all user-setting ConfigMaps:

```
git clone https://github.com/openshift-dev-console/scripts odc-scripts
cd odc-scripts
npm install
./odc user-settings migrate
```

The labels are not consumed at the moment. But they might help other teams like Dev Sandbox to cleanup the right resources in the future.

jerolimov commented:
> Hello Thank you for this PR, have a few questions:
> 
> 1. When looking for resources created by the console, is one label enough to identify? I see that the label `console.openshift.io/user-settings: "true"` is present in all the examples you've mentioned and seems like a good option to use that.

That's enough to identify that this is a user-settings ConfigMap. But this isn't the case for old ConfigMaps.

There is currently no migration code (as part of the console or console-operator) that applies this to old resources.

> 2. Out of the three examples mentioned, it looks like resources created for users in developer sandbox would be of the second example - "developer with UID". If yes, this [PR](https://github.com/codeready-toolchain/member-operator/pull/568) uses the two labels `console.openshift.io/user-settings: "true"  , console.openshift.io/user-settings-uid:` to identify the resources created by console for user, to clean it up, do you think these would be enough to identify them?

Nice that you already picked up this work on your side. Yeah checking for both labels with a uid would be fine. It's also okay to just check for `console.openshift.io/user-settings-uid`.

My idea was that you can check with one label for "all user settings ConfigMaps" and with the other one for the cases that you want remove a specific ConfigMap for a specific user (uid)

> 3. Would it be possible to extend these labels to existing resources as well, maybe like a migration code? It would be easier to identify these resources instead of searching by name. Currently I've divided the code to look for resources by name first and then if not found look by labels - which isn't pretty.

Yep, I shared a Node.js script above that you can run once. We're not sure if we will add this migration also to our operator as part of 4.17 because not all clusters users local OpenShift Users CR... :man_shrugging: 

Wdyt about the node.js migration script? I can create also shell version for adding new labels if this is helpful.

jerolimov commented:
Verified this on a cluster bot (with normal users, no OIDC):

![image](https://github.com/openshift/console/assets/139310/6a8d6a81-db89-4f25-a5b1-672ed080c539)

![image](https://github.com/openshift/console/assets/139310/fed94616-bb32-451e-b245-49240911962d)

Then I manually removed the labels and annotations and run the migration script. It restores the exact same labels and annotations. :heavy_check_mark:

TheRealJon reviewed (approved):
/lgtm

ranakan19 commented:
@jerolimov Thank You for your responses!
 
> There is currently no migration code (as part of the console or console-operator) that applies this to old resources.

Is that something we can consider for future? Not just migration but the ability to clean up these resources by console. I am confident that this does not just impact sandbox. Any customers with openshift clusters irrespective of how they use User CR can have a negative impact on the performance, by leaving roles, rolebindings and configmap without cleaning up. Isn't it?

> Nice that you already picked up this work on your side. Yeah checking for both labels with a uid would be fine. It's also okay to just check for `console.openshift.io/user-settings-uid`.

Gotcha, Thanks! I'll continue with both labels.

> Yep, I shared a Node.js script above that you can run once. We're not sure if we will add this migration also to our operator as part of 4.17 because not all clusters users local OpenShift Users CR... 🤷‍♂️
> Wdyt about the node.js migration script? I can create also shell version for adding new labels if this is helpful.

Yes please! the node.js script looks good but a shell version would be even better. 

Also, since this PR got reverted, is their anyone working on fixing it?
